### ISSUE TITLE
Make Adhoc signing into a Signer module

### ISSUE THREAD

**Issue by neilk** [isignpy/isign#2]
Adhoc signing is effectively a no-op Signer module; refactor and redocument

**Comment by neilk** [comment]
This turned out to not be that simple.

TLDR the "Signer" is being used not only to sign things but also as a source of what you might call the "signing posture" or "signing context". So it has to modify behavior in a number of places based on whether we are "adhoc" signing (identityless signing) or not.

We should probably refactor it so this is not true any more (it's a wart) but this is not a GOAL any more. I added this as a task because it seemed like it could be a useful test of the Signer module concept, but it actually isn't.


Quoting myself from Slack discussion:

----
I originally thought I could refactor “Ad-Hoc Signing” into its own module, and get rid of what looked like an obvious code smell, a number of if signer.is_adhoc() calls throughout the codebase
Here’s what that person was trying to do
So, there’s a special signing context called ad-hoc signing, which is “identityless” signing. There is no signature as we understand it, it just hashes the contents and puts them in a bundle. I don’t know why anyone would be using isign for this. (edited) 
Allegedly, the only people who use this in real apps are Apple, because you need it to run things before any other services are operational.
So they say there are a limited number of hashes which are blessed to run without signing, right in the apple kernel.
(Suddenly generating hash collisions look really interesting)
The other use for it is for the Simulator - all Simulator apps are adhoc signed
anyway, from our perspective the AdHoc signer is just doing no-ops, it returns the empty string as a signature.
BUT
Here’s the other thing; in an adhoc, identityless context, there’s a lot of other things you need to do or not do

Like
- in requirements, return an empty block
- the “teamID” is the empty string and one has to specify version 0x20100 instead of 0x20200
- don’t add the entitlements file to the signature
-    related: don’t fill the identity entitlements slot at all
Since we passed the Signer module around to all those places, the person who wrote in the ad-hoc option just added a test for .is_adhoc(). So effectively the Signer is also doing the job of communicating a sort of configuration.
So the Signer module has the responsibility not just for signing, but also for telling other parts of the system to construct certain things or not. (edited) 
Bugs me a bit. It feels like maybe there’s a missing “signing context”
It seems like it would be way too hard to add that in right now, though.
isinstance(signer, 'AdhocSigner') feels wrong, it really should not be tied to a class hierarchy.
So I guess I’m preserving the .is_adhoc for now :confused:
